File: BlackFireIM/Example/Conversation/View/BFNaviBarIndicatorView.m
//
//  BFNaviBarIndicatorView.m
//  BlackFireIM
//
//  Created by benny wang on 2021/3/18.
//

#import "BFNaviBarIndicatorView.h"
#import "BFHeader.h"
#import "UIColor+BFDarkMode.h"

@implementation BFNaviBarIndicatorView

- (instancetype)initWithFrame:(CGRect)frame
{
    if (self = [super initWithFrame:frame]) {
        [self setupViews];
    }
    return self;
}

- (void)setupViews
{
    _indicator = [[UIActivityIndicatorView alloc]initWithFrame:CGRectMake(0, 0, 20, 20)];
    _indicator.center = CGPointMake(0, NavBar_Height*0.5);
    _indicator.activityIndicatorViewStyle = UIActivityIndicatorViewStyleGray;
    [self addSubview:_indicator];
    
    _label = [[UILabel alloc]init];
    _label.font = [UIFont systemFontOfSize:17];
    _label.textColor = [UIColor d_colorWithColorLight:TText_Color dark:TText_Color_Dark];
    [self addSubview:_label];
}

- (void)setTitle:(NSString *)title
{
    _label.text = title;
    [self updateLayout];
}

- (void)updateLayout
{
    CGSize labelSize = [_label sizeThatFits:CGSizeMake(Screen_Width, NavBar_Height)];
    CGFloat labelY = 0;
    CGFloat labelX = _indicator.hidden ? 0 : (_indicator.frame.origin.x + _indicator.frame.size.width + 5);
    _label.frame = CGRectMake(labelX, labelY, labelSize.width, NavBar_Height);
    self.frame = CGRectMake(0, 0, labelX+labelSize.width+5, NavBar_Height);
}

- (void)startAnimating
{
    [_indicator startAnimating];
}

- (void)stopAnimating
{
    [_indicator stopAnimating];
}

@end

File: Swift/MSIM-Swift/Conversation/BFNaviBarIndicatorView.swift
//
//  BFNaviBarIndicatorView.swift
//  MSIM-Swift
//
//  Created by benny wang on 2021/8/10.
//

import UIKit

class BFNaviBarIndicatorView: UIView {

    var indicator: UIActivityIndicatorView!
    
    var label: UILabel!
    
    func setTitle(title: String) {
        
        label.text = title
        updateLayout()
    }
    
    func startAnimating() {
        indicator.startAnimating()
    }
    
    func stopAnimating() {
        indicator.stopAnimating()
    }
    
    private func updateLayout() {
        
        let labelSize = label.sizeThatFits(CGSize(width: UIScreen.width, height: UIScreen.height))
        let labelY: CGFloat = 0
        let labelX: CGFloat = indicator.isHidden ? 0 : (indicator.left + indicator.width + 5)
        label.frame = CGRect(x: labelX, y: labelY, width: labelSize.width, height: UIScreen.navBarHeight)
        frame = CGRect(x: 0, y: 0, width: labelX + labelSize.width + 5, height: UIScreen.navBarHeight)
    }
    
    override init(frame: CGRect) {
        super.init(frame: frame)
        
        indicator = UIActivityIndicatorView(frame: CGRect(x: 0, y: 0, width: 20, height: 20))
        indicator.center = CGPoint(x: 0, y: UIScreen.navBarHeight * 0.5)
        indicator.style = .medium
        addSubview(indicator)
        
        label = UILabel()
        label.font = .systemFont(ofSize: 17)
        label.textColor = UIColor.d_color(light: MSMcros.TText_Color, dark: MSMcros.TText_Color_Dark)
        addSubview(label)
    }
    
    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
}
